src/sync/sync_session.cpp

```diff
@@ -404,6 +404,54 @@ void SyncSession::create_sync_session()
         }
     };
     m_session->set_sync_transact_callback(std::move(wrapped_callback));
+
+    // Set up the wrapped progress handler callback
+    auto wrapped_progress_handler = [this, weak_self](uint_fast64_t downloaded,
+                                                      uint_fast64_t downloadable,
+                                                      uint_fast64_t uploaded,
+                                                      uint_fast64_t uploadable) {
+        auto self = weak_self.lock();
+        if (!self) {
+            // An update was delivered after the session it relates to was destroyed.
+            // There's nothing useful we can do with it.
+            return;
+        }
+        std::vector<std::function<void()>> invocations;
+        {
+            std::lock_guard<std::mutex> lock(m_progress_notifier_mutex);
+            m_current_uploadable = uploadable;
+            m_current_downloadable = downloadable;
+            m_current_downloaded = downloaded;
+            m_current_uploaded = uploaded;
+            for (auto& pair : m_notifiers) {
+                auto& package = pair.second;
+                uint64_t transferred = (package.direction == NotifierType::download ? downloaded : uploaded);
+                uint64_t transferrable = (package.direction == NotifierType::download ? downloadable : uploadable);
+                if (package.is_streaming) {
+                    invocations.emplace_back([=, notifier=package.notifier](){
+                        notifier(transferred, transferrable);
+                    });
+                } else {
+                    if (!package.captured_transferrable) {
+                        package.captured_transferrable = transferrable;
+                    }
+                    invocations.emplace_back([=, notifier=package.notifier,
+                                              transferrable=package.captured_transferrable](){
+                        notifier(transferred, transferrable);
+                    });
+                    if (package.captured_transferrable <= transferred) {
+                        // A notifier is expired if the captured max is no longer greater than the current value.
+                        m_notifiers.erase(pair.first);
+                    }
+                }
+            }
+        }
+        for (auto& invocation : invocations) {
+            // Run the notifiers only after we've released the lock.
```

Comment:
I'd suggest moving this comment out of the loop since it's about the loop itself rather than the body of it.